pricegraph/src/num.rs

```diff
@@ -4,9 +4,11 @@ use primitive_types::U256;
 use std::cmp;
 use std::f64;
 
-/// The maximum rounding error, used for asserting that amounts and balances
-/// remain coherent for tests and when built in `debug` profile.
-pub const MAX_ROUNDING_ERROR: f64 = 0.1;
+/// The maximum rounding error for the specified amount, used for asserting that
+/// amounts and balances remain coherent for tests and for `debug` profile.
+pub fn max_rounding_error(quantity: f64) -> f64 {
+    quantity * f64::EPSILON
```

Comment:
I'd like to see reasoning for this formular. Is this a standard thing to do?

Reply:
> I'd like to see reasoning for this formula

The idea is to find the value of the least significant bit of the mantissa - since we want to **at most** be off by 1 bit.

So, the `f64::EPISILON` value is the smallest value that can be added to 1.0. That means no other real values in the range`(1.0, 1.0 + f64::EPSILON)` can be represented by floating point numbers.

So given a function like this:
```rust
fn h() {
    println!("s eeeeeeeeeee mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm");
}
fn d(a: f64) {
    let a = a.to_bits();
    println!("{:b} {:011b} {:052b}", a >> 63, (a >> 52) & 0x7ff, a & 0x000fffffffffffff);
}
```

We can check how the mantissa and exponent of `1.0` and `1.0 + f64::EPSILON` look like:
```rust
h();
d(1.0);
d(1.0 + f64::EPSILON);
```

Outputs
```
s eeeeeeeeeee mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm
0 01111111111 0000000000000000000000000000000000000000000000000000
s eeeeeeeeeee mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm
0 01111111111 0000000000000000000000000000000000000000000000000001
```

So basically adding `f64::EPSILON` ads one to the least significant bit of the mantissa.

If you want to apply this to larger amounts, you can add `amount + amount * f64::EPSILON` to much the same effect. Using numbers from the fuzzy testing example.

```rust
h();
let a = 83_798_276_971_421_254_262_445_676_335_662_107_162.0f64;
d(a);
d(a + a * f64::EPSILON);
```

Outputs:
```
s eeeeeeeeeee mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm
0 10001111100 1111100001010111101100000101000001010110011001000001
0 10001111100 1111100001010111101100000101000001010110011001000011
```

Here, it rounds by 2 bits! This is actually because of rounding, since the `1.mantissa > 1.5`.

So if we change it a little by removing all mantissa bits, we can make the spread a little tighter:
```rust
h();
let a = 83_798_276_971_421_254_262_445_676_335_662_107_162.0f64;
d(a);
d(a + a * (f64::from_bits(a.to_bits() & 0xfff0_0000_0000_0000) * f64::EPSILON));
```

Which outputs what we want:
```
s eeeeeeeeeee mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm
0 10001111100 1111100001010111101100000101000001010110011001000001
0 10001111100 1111100001010111101100000101000001010110011001000010
```

We can even check that this works on exponent boundries
```rust
fn r(a: f64) -> f64 {
    f64::from_bits(a.to_bits() & 0xfff0_0000_0000_0000) * f64::EPSILON
}

let a = f64::from_bits(0b0_10001111100_1111111111111111111111111111111111111111111111111111);
let b = f64::from_bits(0b0_10001111101_0000000000000000000000000000000000000000000000000000);

d(a);
d(a + r(a));
d(b);
d(r(a));
d(b - a);
```

Which outputs:
```
s eeeeeeeeeee mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm
0 10001111100 1111111111111111111111111111111111111111111111111111
0 10001111101 0000000000000000000000000000000000000000000000000000
0 10001111101 0000000000000000000000000000000000000000000000000000
0 10001001000 0000000000000000000000000000000000000000000000000000
0 10001001000 0000000000000000000000000000000000000000000000000000
```

I'm going to modify the code to use this new rounding error formula.

Its a good thing I had to write this out, :sweat_smile: